File: scripts/py/audio_words_heard.py
"""What a recording sounds like it says, to something that was never shown its words.

This is the opposite choice from the aligner beside it, and the two are meant to
be read together.  An aligner is handed the words and can only decide where they
fall; it therefore cannot disagree with the text, and its own score is no guard,
because the score is worked out by the very model whose confusion is in doubt.
A transcriber is told nothing, so what comes back is an independent hearing, and
how much of it matches the written text is a judgement the aligner is unable to
make about itself.

Measured on a sung psalm: the aligner's confidence was 0.367 against the psalm's
own words and 0.25 against a different psalm's, which separates nothing, while
this hearing matched 0.968 of the right words and 0.191 of the wrong ones.  The
floor the aligner ships was measured on speech and does not survive the move to
singing; an independent hearing does.

The model is Whisper as rebuilt for CTranslate2, published under the MIT licence,
weights and code alike.  It is chosen over the alignment models for this job
because it was trained on ordinary recorded audio rather than on people reading
books aloud, so singing is inside what it has heard before rather than outside
it.  Its word times are not used for placing lines: they run about four tenths of
a second early and stay about a quarter of a second out once that is taken off,
where the aligner is inside a tenth.  It is here to say what was sung, not when.

The times it does give are still worth carrying, because two independent readings
of the same line can be compared: where they disagree by more than about a third
of a second, one of them is wrong, and on a psalm whose right times were known
that flagged thirteen lines of thirty two and caught all seven that were actually
misplaced.

Takes the path of a JSON file holding {"audio": <path>, "model": <a whisper size
name>} and prints one JSON line holding every word heard with the second it
begins and the second it ends, counted from the start of the file, each with the
model's own probability for it.
"""

import json
import sys

from faster_whisper import WhisperModel


def main(args_path):
    with open(args_path) as handle:
        args = json.load(handle)

    model = WhisperModel(args["model"], device="cpu", compute_type="int8")
    segments, _info = model.transcribe(
        args["audio"],
        word_timestamps=True,
        language="en",
        vad_filter=False,
    )

    words = []
    for segment in segments:
        for word in segment.words or []:
            words.append(
                {
                    "word": word.word.strip(),
                    "start": round(word.start, 3),
                    "end": round(word.end, 3),
                    "score": round(word.probability, 3),
                }
            )

    print(json.dumps({"words": words, "count": len(words)}, ensure_ascii=False))


if __name__ == "__main__":
    main(sys.argv[1])

File: js/audio_words_heard.mjs
import { fn_name } from "./fn_name.mjs";
import { arguments_assert } from "./arguments_assert.mjs";
import { audio_words_heard_model } from "./audio_words_heard_model.mjs";
import { py_script_speech_json_report } from "./py_script_speech_json_report.mjs";
import { equal } from "./equal.mjs";
import { property_get } from "./property_get.mjs";
export async function audio_words_heard(path_audio) {
  "$plain path_audio";
  "Every word a recording sounds like it says, to a listener that was never shown its words, each with the second it begins and the second it ends.";
  "★ IT IS TOLD NOTHING ON PURPOSE, AND THAT IS THE WHOLE VALUE OF IT. Its twin lays a known text onto the sound and so can never disagree with that text, and the confidence the twin reports is worked out by the same model whose confusion is the thing in doubt. This one has no text to agree with, so how much of what it heard matches what is written is a judgement neither the aligner nor its score can make.";
  "★ THE TIMES THAT COME BACK ARE FOR COMPARING, NOT FOR PLACING. Measured against known times on a sung psalm they run about four tenths of a second early and remain about a quarter of a second out once that shift is taken off, where the aligner is inside a tenth. So a caller should take its lines from the aligner and use these to ask whether the aligner was right - where two independent readings of one line disagree by more than about a third of a second, one of them is wrong.";
  "It hears the whole file in one call because there is nothing to divide: a transcriber decides for itself where one stretch of speech ends, and a caller that cut the file into pieces first would be handing it a decision it makes better than they can.";
  arguments_assert(arguments, 1);
  let model = audio_words_heard_model();
  let args = {
    audio: path_audio,
    model: model,
  };
  let reported = await py_script_speech_json_report(
    fn_name("audio_words_heard"),
    args,
  );
  if (equal(reported, null)) {
    return null;
  }
  let words = property_get(reported, "words");
  return words;
}
